lib/picotls.c

```
@@ -2478,6 +2488,18 @@ static int client_handle_encrypted_extensions(ptls_t *tls, ptls_iovec_t message,
             }
             skip_early_data = 0;
             break;
+        case PTLS_EXTENSION_TYPE_SERVER_CERTIFICATE_TYPE:
+            if (end - src != 1) {
+                ret = PTLS_ALERT_DECODE_ERROR;
+                goto Exit;
+            }
+            if ((tls->ctx->cert0_is_raw_certificate && *src != PTLS_CERTIFICATE_TYPE_RAW_PUBLIC_KEY) &&
+                *src != PTLS_CERTIFICATE_TYPE_X509) {
```

What is the expected behavior when the client sends ServerCertType(RAW) and the server responds with ServerCertType(X509)? I tend to think that we should raise an UNSUPPORTED_CERTIFICATE error, because X509 is not what the client has requested, and because the verify_certificate callback cannot handle unexpected type of certificate. However, it seems to me that the code tries to continue the handshake in such case.

I tend to believe that this condition should be `ctx->cert0_is_raw_certificate == (server_offered_type == PTLS_CERTIFICATE_TYPE_RAW_PUBLIC_KEY)`.

In addition, we have to run the same check when the server does not send the extension (in which case server is trying to send an X.509 certificate). Therefore, what we might want to do is memorize the type being offered by the server (which reflects either the omission of the extension or the value carried by the extension), then verify it after decoding all the extensions.